test/controllers/tasks_controller_test.rb

```diff
@@ -79,48 +72,172 @@
-      # Your code here
+      updated_task_data = {
+        task: {
+          name: "Walk dog",
+          description: "Walk the dog",
+          completion_date: nil,
+        },
+      }
+      task_url = "/tasks/" + "#{task.id}"
+      
+      patch task_url, params: updated_task_data
+      
+      expect(Task.count).must_equal 1
+      must_redirect_to task_path(id: task.id)
+      expect(Task.find_by(id: task.id).name).must_equal "Walk dog"
     end
-
+    
     it "will redirect to the root page if given an invalid id" do
-      # Your code here
+      patch '/tasks/0', params: {
+        task: {
+          name: "Walk dog",
+          description: "Walk the dog",
+          completion_date: nil,
+        },
+      }
+      must_redirect_to root_path
+    end
+    
+    it "will not update if given invalid params" do
+      new_task = Task.create(
+        name: "Make dinner", 
+        description: "Find something edible in the fridge",
+        completion_date: nil
+      )
+      
+      expect {
+        patch task_path(new_task_id), params: {}
+      }.must_raise 
+      expect(new_task.name).must_equal "Make dinner"
+      expect(new_task.description).must_equal "Find something edible in the fridge"
+      expect(new_task.completion_date).must_be_nil
     end
   end
 
   # Complete these tests for Wave 4
   describe "destroy" do
-    # Your tests go here
+    it 'successfully deletes a task and redirects to the root path' do
+      new_task = Task.create(
+        name: "Make dinner", 
+        description: "Find something edible in the fridge",
+        completion_date: nil
+      )
+      new_task_id = new_task.id
+
+      expect{
+        delete task_path(new_task_id)
+      }.must_differ "Task.count", -1
+
+      must_redirect_to root_path
+    end
+
+    it "redirects to the tasks path and does not delete any tasks if the task does not exist" do
+      Task.destroy_all
+      new_task = Task.create(
+        name: "Make dinner", 
+        description: "Find something edible in the fridge",
+        completion_date: nil
+      )
+      invalid_task_id = 600
+
+      expect {
+        delete task_path(invalid_task_id)
+      }.must_differ "Task.count", 0
+      expect(new_task.id).wont_equal invalid_task_id
+
+      must_redirect_to tasks_path
+    end
+
+    it "redirects to the tasks path and does not delete any tasks if the task has already been deleted" do
+      new_task = Task.create(
+        name: "Make dinner", 
+        description: "Find something edible in the fridge",
+        completion_date: nil
+      )
+      new_task_id = new_task.id
+      Task.destroy_all
+
+      expect {
+        delete task_path(new_task_id)
+      }.must_differ "Task.count", 0
 
+      must_redirect_to tasks_path
+    end
   end
 
   # Complete for Wave 4
   describe "toggle_complete" do
-    # Your tests go here
+    it "successfully updates completion_date to today's date if nil and redirects to the tasks path" do
```

Comment:
Nicely done on all of these `toggle_complete` tests! They are thorough and well-tested!